README.md

```diff
@@ -91,6 +95,79 @@ an apostrophe <<'>> will be appended to the identifier with different value.
 
 The data reference identifiers will be prepended with `&` in the report
 
+### Assertions
+
+Specially-formatted comments can be used to put constraints on values on top of the stack: `// bsst-assert:` and `// bsst-assert-size:`. The difference is that the former puts constraints on the value itself, while the latter constraints the
+data size instead of value
+
+The expressibility of these assertions are limited, as their primary purpose is to help the solver by reducing the range of values to be considered.
+
+For the value on top of the stack constrained via assertion, B'SST will check
+if the value can happen to be outside the range defined by the assertion
+expression. If it can, the currently analyzed execution path will be deemed
+as failed, and in the report the failure will be shown as
+`assertion failed at line <N>` or `check_assertion_at_line_<N>` where `<N>`
+would be the line at which the failing `// bsst-assert:` comment is at
+
+After the check successfully passed, the value will be assumed to be constrained
+by the assertion expression.
+
+But if the assertion is applied to the data placeholder, no check will be
+performed, and the value of the data placeholder will simply be assumed to be
+restricted by the assertion expression.
```

Comment:
One thought: It might be better to split this behaviour into two comments: `bsst-assert` that gives an error if the value is outside the range specified (or can't be proven to be inside it), and `bsst-assume` that treats the expression as if it were proven (and can only be applied to fresh data placeholders?)

Reply:
I have thought about splitting it between assert/assume, but decided to start with just assert for both behaviors because it is simpler.

I will consider separating into assert and assume.

What do you mean by "fresh" data placeholder ? Like, the first time the data placeholder is seen in the script ? But what if data placeholder is placed in two branches of `IF` ? Because the constraints are applied alongside with per-op analysis, the second instance of data placeholder (on the false branch execution path) might not get its constraints.

Maybe I should just make `bsst-assume` take the name of the data placeholder as first argument, and then have constraints applied on the first encounter of that data placeholder. So you would have something like

```
// bsst-assume: $a <=3 12
...
DUP ADD
$a SUB
// bsst-assert: != 0
...
```
And then you could have `bsst-assert` after data placeholder, too - because constraints put upon it by other opcodes might restrict its value enough for assertion to pass

Reply:
> But what if data placeholder is placed in two branches of IF ? Because the constraints are applied alongside with per-op analysis, the second instance of data placeholder (on the false branch execution path) might not get its constraints.

Well, actually this is true for current `bsst-assert` behavior, too. So this is another reason to make `bsst-assume:` to take data placeholder name as argument rather than work with top of the stack